File: 0031.NextPermutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        if len(nums) <= 1:
            return
        i = len(nums)-2
        while i >= 0 and nums[i] >= nums[i+1]:
            i -= 1
        if i >= 0:
            j = i+1
            while j < len(nums)-1 and nums[j+1] > nums[i]:
                j += 1
            nums[i], nums[j] = nums[j], nums[i]
            nums[i+1:] = reversed(nums[i+1:])
            #nums[i+1:] = sorted(nums[i+1:])
        else:
            nums.reverse()

File: 0031.NextPermutation.go
func nextPermutation(nums []int)  {
    n := len(nums)
    i := n - 2
    for ; i >= 0 && nums[i] >= nums[i+1]; i-- {}
    if i < 0 {
        reverse(nums, 0, n-1)
        return
    }
    j := n - 1
    for ; j > i && nums[j] <= nums[i]; j-- {}
    nums[i], nums[j] = nums[j], nums[i]
    reverse(nums, i+1, n-1)
}

func reverse(nums []int, left, right int) {
    for ; left < right; left, right = left+1, right-1 {
        nums[left], nums[right] = nums[right], nums[left]
    }
}